File: native-ios/BeerNative/Sources/RpgModels.swift
import Foundation

struct RpgState: Decodable {
    var enabled: Bool?
    var ui: Bool?
    var allowed: Bool?
    var profile: RpgProfile?
    var quests: RpgQuests?
    var badges: [RpgBadge]?
    var nextBadges: [RpgBadge]?
    var atlas: RpgAtlas?
    var classes: [RpgClassInfo]?
    var classAffinity: [String: Int]?
    var phrase: String?

    enum CodingKeys: String, CodingKey {
        case enabled, ui, allowed, profile, quests, badges, atlas, classes, phrase
        case nextBadges = "next_badges"
        case classAffinity = "class_affinity"
    }

    var active: Bool {
        enabled == true && ui == true && profile != nil
    }
}

struct RpgProfile: Decodable {
    var username: String?
    var level: Int?
    var xp: Int?
    var title: String?
    var progressPct: Double?
    var xpToNext: Int?
    var xpIntoLevel: Int?
    var xpLevelStart: Int?
    var xpLevelNext: Int?
    var streakDays: Int?
    var dailyXp: Int?
    var dailySoftCap: Int?
    var classKey: String?
    var classInfo: RpgClassInfo?
    var beerMaster: Bool?
    var prestige: RpgPrestige?
    var titleBand: RpgTitleBand?

    enum CodingKeys: String, CodingKey {
        case username, level, xp, title, prestige
        case progressPct = "progress_pct"
        case xpToNext = "xp_to_next"
        case xpIntoLevel = "xp_into_level"
        case xpLevelStart = "xp_level_start"
        case xpLevelNext = "xp_level_next"
        case streakDays = "streak_days"
        case dailyXp = "daily_xp"
        case dailySoftCap = "daily_soft_cap"
        case classKey = "class"
        case classInfo = "class_info"
        case beerMaster = "beer_master"
        case titleBand = "title_band"
    }

    var displayIcon: String {
        if beerMaster == true { return prestige?.icon ?? "👑" }
        return classInfo?.icon ?? "🍺"
    }
}

struct RpgClassInfo: Decodable, Identifiable {
    var key: String?
    var name: String?
    var icon: String?
    var blurb: String?
    var whenText: String?
    var special: String?
    var id: String { key ?? name ?? UUID().uuidString }

    enum CodingKeys: String, CodingKey {
        case key, name, icon, blurb, special
        case whenText = "when"
    }
}

struct RpgPrestige: Decodable {
    var key: String?
    var icon: String?
    var ribbon: String?
    var tagline: String?
    var blurb: String?
}

struct RpgTitleBand: Decodable {
    var name: String?
    var fromLevel: Int?
    var to: Int?
    enum CodingKeys: String, CodingKey {
        case name, to
        case fromLevel = "from"
    }
}

struct RpgQuests: Decodable {
    var active: [RpgQuest]?
    var doneToday: [RpgQuest]?
    var doneWeekly: [RpgQuest]?
    enum CodingKeys: String, CodingKey {
        case active
        case doneToday = "done_today"
        case doneWeekly = "done_weekly"
    }
}

struct RpgQuest: Decodable, Identifiable {
    var key: String?
    var kind: String?
    var title: String?
    var description: String?
    var progress: Int?
    var target: Int?
    var status: String?
    var rewardXp: Int?
    var id: String { key ?? title ?? UUID().uuidString }
    enum CodingKeys: String, CodingKey {
        case key, kind, title, description, progress, target, status
        case rewardXp = "reward_xp"
    }
}

struct RpgBadge: Decodable, Identifiable {
    var key: String?
    var name: String?
    var icon: String?
    var rarity: String?
    var lore: String?
    var hint: String?
    var earned: Bool?
    var earnedAt: String?
    var progress: Int?
    var target: Int?
    var remaining: Int?
    var id: String { key ?? name ?? UUID().uuidString }
    enum CodingKeys: String, CodingKey {
        case key, name, icon, rarity, lore, hint, earned, progress, target, remaining
        case earnedAt = "earned_at"
    }
}

struct RpgAtlas: Decodable {
    var stylesCount: Int?
    var hopsCount: Int?
    var breweriesCount: Int?
    var photos: Int?
    var totalCheckins: Int?
    var styles: [String]?
    enum CodingKeys: String, CodingKey {
        case photos, styles
        case stylesCount = "styles_count"
        case hopsCount = "hops_count"
        case breweriesCount = "breweries_count"
        case totalCheckins = "total_checkins"
    }
}

struct RpgLoot: Decodable {
    var xpGained: Int?
    var xp: Int?
    var level: Int?
    var levelUp: Bool?
    var oldLevel: Int?
    var title: String?
    var progressPct: Double?
    var xpToNext: Int?
    var phrase: String?
    var phraseLevelUp: String?
    var badgesEarned: [RpgBadge]?
    var questsCompleted: [RpgQuest]?
    var nextBadges: [RpgBadge]?
    var streakDays: Int?
    enum CodingKeys: String, CodingKey {
        case xp, level, title, phrase
        case xpGained = "xp_gained"
        case levelUp = "level_up"
        case oldLevel = "old_level"
        case progressPct = "progress_pct"
        case xpToNext = "xp_to_next"
        case phraseLevelUp = "phrase_level_up"
        case badgesEarned = "badges_earned"
        case questsCompleted = "quests_completed"
        case nextBadges = "next_badges"
        case streakDays = "streak_days"
    }
}

func rarityLabelFr(_ r: String?) -> String {
    switch (r ?? "common").lowercased() {
    case "legendary": return "Légendaire"
    case "epic": return "Épique"
    case "rare": return "Rare"
    default: return "Commun"
    }
}

File: native-android/app/src/main/java/fr/eiter/plexibeer/RpgModels.kt
package fr.eiter.plexibeer

import com.google.gson.annotations.SerializedName
import com.google.gson.JsonElement

/** GET /api/rpg/me */
data class RpgState(
    val enabled: Boolean = false,
    val ui: Boolean = false,
    val allowed: Boolean = false,
    val profile: RpgProfile? = null,
    val quests: RpgQuests? = null,
    val badges: List<RpgBadge> = emptyList(),
    @SerializedName("next_badges") val nextBadges: List<RpgBadge> = emptyList(),
    val atlas: RpgAtlas? = null,
    val classes: List<RpgClassInfo> = emptyList(),
    @SerializedName("class_affinity") val classAffinity: Map<String, Int>? = null,
    val phrase: String? = null
)

data class RpgProfile(
    val username: String? = null,
    val level: Int = 1,
    val xp: Int = 0,
    val title: String? = null,
    @SerializedName("progress_pct") val progressPct: Double = 0.0,
    @SerializedName("xp_to_next") val xpToNext: Int? = null,
    @SerializedName("xp_into_level") val xpIntoLevel: Int? = null,
    @SerializedName("xp_level_start") val xpLevelStart: Int? = null,
    @SerializedName("xp_level_next") val xpLevelNext: Int? = null,
    @SerializedName("streak_days") val streakDays: Int = 0,
    @SerializedName("daily_xp") val dailyXp: Int = 0,
    @SerializedName("daily_soft_cap") val dailySoftCap: Int = 100,
    @SerializedName("class") val classKey: String? = null,
    @SerializedName("class_info") val classInfo: RpgClassInfo? = null,
    @SerializedName("beer_master") val beerMaster: Boolean = false,
    val prestige: RpgPrestige? = null,
    @SerializedName("title_band") val titleBand: RpgTitleBand? = null,
    @SerializedName("intro_seen") val introSeen: Boolean = false
)

data class RpgClassInfo(
    val key: String? = null,
    val name: String? = null,
    val icon: String? = null,
    val blurb: String? = null,
    @SerializedName("when") val whenText: String? = null,
    val special: String? = null
)

data class RpgPrestige(
    val key: String? = null,
    val icon: String? = null,
    val ribbon: String? = null,
    val tagline: String? = null,
    val blurb: String? = null
)

data class RpgTitleBand(
    val name: String? = null,
    @SerializedName("from") val fromLevel: Int? = null,
    val to: Int? = null
)

data class RpgQuests(
    val active: List<RpgQuest> = emptyList(),
    @SerializedName("done_today") val doneToday: List<RpgQuest> = emptyList(),
    @SerializedName("done_weekly") val doneWeekly: List<RpgQuest> = emptyList()
)

data class RpgQuest(
    val key: String? = null,
    val kind: String? = null,
    val title: String? = null,
    val description: String? = null,
    val progress: Int = 0,
    val target: Int = 1,
    val status: String? = null,
    @SerializedName("reward_xp") val rewardXp: Int = 0
)

data class RpgBadge(
    val key: String? = null,
    val name: String? = null,
    val icon: String? = null,
    val rarity: String? = null,
    val lore: String? = null,
    val hint: String? = null,
    val earned: Boolean = false,
    @SerializedName("earned_at") val earnedAt: String? = null,
    val progress: Int = 0,
    val target: Int = 1,
    val remaining: Int? = null
)

data class RpgAtlas(
    @SerializedName("styles_count") val stylesCount: Int = 0,
    @SerializedName("hops_count") val hopsCount: Int = 0,
    @SerializedName("breweries_count") val breweriesCount: Int = 0,
    val photos: Int = 0,
    @SerializedName("total_checkins") val totalCheckins: Int = 0,
    val styles: List<String>? = null
)

/** Bloc `rpg` renvoyé par POST /api/checkins */
data class RpgLoot(
    @SerializedName("xp_gained") val xpGained: Int = 0,
    val xp: Int = 0,
    val level: Int = 1,
    @SerializedName("level_up") val levelUp: Boolean = false,
    @SerializedName("old_level") val oldLevel: Int? = null,
    val title: String? = null,
    @SerializedName("progress_pct") val progressPct: Double = 0.0,
    @SerializedName("xp_to_next") val xpToNext: Int? = null,
    val phrase: String? = null,
    @SerializedName("phrase_level_up") val phraseLevelUp: String? = null,
    @SerializedName("badges_earned") val badgesEarned: List<RpgBadge> = emptyList(),
    @SerializedName("quests_completed") val questsCompleted: List<RpgQuest> = emptyList(),
    @SerializedName("next_badges") val nextBadges: List<RpgBadge> = emptyList(),
    @SerializedName("streak_days") val streakDays: Int? = null,
    /** breakdown items may be heterogeneous — keep as JsonElement list if needed */
    val breakdown: List<Map<String, JsonElement>>? = null
)

fun RpgProfile.displayIcon(): String {
    if (beerMaster) return prestige?.icon ?: "👑"
    return classInfo?.icon ?: "🍺"
}

fun rarityLabelFr(r: String?): String = when ((r ?: "common").lowercase()) {
    "legendary" -> "Légendaire"
    "epic" -> "Épique"
    "rare" -> "Rare"
    else -> "Commun"
}
